libctru/source/services/nim.c

```diff
@@ -0,0 +1,254 @@
+#include <string.h>
+#include <stdlib.h>
+#include <malloc.h>
+#include <3ds/types.h>
+#include <3ds/result.h>
+#include <3ds/svc.h>
+#include <3ds/srv.h>
+#include <3ds/synchronization.h>
+#include <3ds/services/nim.h>
+#include <3ds/ipc.h>
+#include <3ds/util/utf.h>
+
+static Handle nimsHandle;
+static int nimsRefCount;
+
+static Result NIMS_GetInitializeResult(void)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
+	cmdbuf[0] = IPC_MakeHeader(0x3f, 0, 0); // 0x003f0000
+
+	if (R_FAILED(ret = svcSendSyncRequest(nimsHandle))) return ret;
+
+	return (Result)cmdbuf[1];
+}
+
+static Result NIMS_RegisterSelf(void)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
+	cmdbuf[0] = IPC_MakeHeader(0x3c, 0, 2); // 0x003c0002
+	cmdbuf[1] = IPC_Desc_CurProcessHandle();
+	cmdbuf[2] = IPC_Desc_MoveHandles(1);
+
+	if (R_FAILED(ret = svcSendSyncRequest(nimsHandle))) return ret;
+
+	return (Result)cmdbuf[1];
+}
+
+static Result NIMS_SetTIN(void)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	static char attr[] = "TIN";
+	static char TIN[] = "56789"; // Obtained from eShop
+
+	cmdbuf[0] = IPC_MakeHeader(0xb, 2, 4); // 0x000b0084
+	cmdbuf[1] = sizeof(attr);
+	cmdbuf[2] = sizeof(TIN);
+	cmdbuf[3] = IPC_Desc_StaticBuffer(cmdbuf[1], 0);
+	cmdbuf[4] = (u32)attr;
+	cmdbuf[5] = IPC_Desc_StaticBuffer(cmdbuf[2], 1);
+	cmdbuf[6] = (u32)TIN;
+
+	if (R_FAILED(ret = svcSendSyncRequest(nimsHandle))) return ret;
+
+	return (Result)cmdbuf[1];
+}
+
+static Result NIMS_ConnectNoTicketDownload(void)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	static u8 buffer[0x20000];
```

Comment:
Why is this hardcoded? Also, this is not thread safe. If you really do not want to allow the user to pass their own buffer, this should be marked as thread-local (refer to the sdmc or romfs code for an example).